server/datastore/s3/carves.go

```diff
@@ -0,0 +1,236 @@
+// results are wrapped into the s3.CompletedPart struct
+func (d *Datastore) listCompletedParts(objectKey, uploadID string) ([]*s3.CompletedPart, error) {
+	var res []*s3.CompletedPart
+	var partMarker int64
+	for {
+		parts, err := d.s3client.ListParts(&s3.ListPartsInput{
+			Bucket:           &d.bucket,
+			Key:              &objectKey,
+			UploadId:         &uploadID,
+			PartNumberMarker: &partMarker,
+		})
+		if err != nil {
+			return res, err
+		}
+		for _, p := range parts.Parts {
+			res = append(res, &s3.CompletedPart{
+				ETag:       p.ETag,
+				PartNumber: p.PartNumber,
+			})
+		}
+		if !*parts.IsTruncated {
+			break
+		}
+		partMarker = *parts.NextPartNumberMarker
+	}
+	return res, nil
+}
+
+// NewBlock uploads a new block for a specific carve
+func (d *Datastore) NewBlock(metadata *kolide.CarveMetadata, blockID int64, data []byte) error {
+	objectKey := d.generateS3Key(metadata)
+	partNumber := blockID + 1 // PartNumber is 1-indexed
+	_, err := d.s3client.UploadPart(&s3.UploadPartInput{
+		Body:       bytes.NewReader(data),
+		Bucket:     &d.bucket,
+		Key:        &objectKey,
+		PartNumber: &partNumber,
+		UploadId:   &metadata.SessionId,
+	})
+	if err != nil {
+		return errors.Wrap(err, "s3 multipart carve upload")
+	}
+	if metadata.MaxBlock < blockID {
+		metadata.MaxBlock = blockID
+		err = d.UpdateCarve(metadata)
+		if err != nil {
+			return err
+		}
+	}
+	if blockID >= metadata.BlockCount-1 {
+		// The last block was reached, multipart upload can be completed
+		parts, err := d.listCompletedParts(objectKey, metadata.SessionId)
+		if err != nil {
+			return errors.Wrap(err, "s3 multipart carve upload")
+		}
+		_, err = d.s3client.CompleteMultipartUpload(&s3.CompleteMultipartUploadInput{
+			Bucket:          &d.bucket,
+			Key:             &objectKey,
+			UploadId:        &metadata.SessionId,
+			MultipartUpload: &s3.CompletedMultipartUpload{Parts: parts},
+		})
+		if err != nil {
+			return errors.Wrap(err, "s3 multipart carve upload")
+		}
+	}
+	return nil
+}
+
+// GetBlock returns a block of data for a carve
+func (d *Datastore) GetBlock(metadata *kolide.CarveMetadata, blockID int64) ([]byte, error) {
+	objectKey := d.generateS3Key(metadata)
+	// blockID is 0-indexed and sequential so can be perfectly used for evaluating ranges
+	// range extremes are inclusive as for RFC-2616 (section 14.35)
+	// no need to cap the rangeEnd to the carve size as S3 will do that by itself
+	rangeStart := blockID * metadata.BlockSize
+	rangeString := fmt.Sprintf("bytes=%d-%d", rangeStart, rangeStart+metadata.BlockSize-1)
+	res, err := d.s3client.GetObject(&s3.GetObjectInput{
+		Bucket: &d.bucket,
+		Key:    &objectKey,
+		Range:  &rangeString,
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == s3.ErrCodeNoSuchKey {
+			// The carve does not exists in S3, mark expired
+			metadata.Expired = true
+			d.UpdateCarve(metadata)
```

Comment:
Please check for error here.

Reply:
Can do, but I honestly wouldn't really know where to route it. I feel like the "not found" error returned from S3 should have priority in this case (even if the DB update fails). Any suggestion? Should I just log it?